yamui-screensaverd.c

```diff
@@ -0,0 +1,361 @@
+
+/* Check for input device type. Returns 0 if button or touchscreen. */
+static int
+check_device_type(int fd)
+{
+	int i;
+	unsigned long bits[EV_MAX][NBITS(KEY_MAX)];
+
+	memset(bits, '\0', sizeof(bits));
+	if (ioctl(fd, EVIOCGBIT(0, EV_MAX), bits[0]) == -1) {
+		errorf("ioctl(, EVIOCGBIT(0, ), ) error on event device");
+		return -1;
+	}
+
+	for (i = 0; i < EV_MAX; i++)
+		if (BIT(bits[0], EV_ABS)) {
+			if (ioctl(fd, EVIOCGBIT(EV_ABS, KEY_MAX),
+				  bits[EV_ABS]) == -1) {
+				errorf("ioctl(, EVIOCGBIT(EV_ABS, ), ) error "
+				       "on event device");
+				return -1;
+			}
+
+			if (BIT(bits[EV_ABS], ABS_MT_POSITION_X) &&
+			    BIT(bits[EV_ABS], ABS_MT_POSITION_Y)) {
+				debugf("Device supports multi-touch events.");
+				return 0;
+			}
+		} else if (BIT(bits[0], EV_KEY)) {
+			if (ioctl(fd, EVIOCGBIT(EV_KEY, KEY_MAX),
+				  bits[EV_KEY]) == -1) {
+				errorf("ioctl(, EVIOCGBIT(EV_KEY, ), ) error "
+				       "on event device");
+				return -1;
+			}
+
+			if (BIT(bits[EV_KEY], KEY_POWER)	||
+			    BIT(bits[EV_KEY], KEY_VOLUMEDOWN)	||
+			    BIT(bits[EV_KEY], KEY_VOLUMEUP)	||
+			    BIT(bits[EV_KEY], KEY_OK)		||
+			    BIT(bits[EV_KEY], KEY_ENTER)) {
+				debugf("Device supports needed key events.");
+				return 0;
+			}
+		}
+
+	debugf("Skipping unsupported device.");
+	return -1;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static void
+close_fds(int fds[], int num)
+{
+	int i;
+
+	for (i = 0; i < num; i++)
+		close(fds[i]);
+}
+
+/* ------------------------------------------------------------------------ */
+
+/* Open all /dev/input/event* files. */
+static int
+open_fds(int fds[], int *num, int max_num)
+{
+	DIR *dir;
+	struct dirent *d;
+
+	if (!(dir = opendir(DEV_INPUT_DIR))) {
+		errorf("Can't open directory %s", DEV_INPUT_DIR);
+		return -1;
+	}
+
+	while ((d = readdir(dir))) {
+		char name[256];
+
+		if (strncmp(d->d_name, EVENT_PREFIX, strlen(EVENT_PREFIX)))
+			continue; /* Not /dev/input/event* file */
+
+		snprintf(name, sizeof(name), "%s/%s", DEV_INPUT_DIR,
+			 d->d_name);
+		debugf("Processing input ivents file %s", name);
+		if ((fds[*num] = open(name, O_RDONLY)) == -1) {
+			errorf("Can't open input device %s", name);
+			continue;
+		}
+
+		if (check_device_type(fds[*num]) == -1) {
+			close(fds[*num]);
+			continue;
+		}
+
+		if (++*num >= max_num)
+			break;
+	}
+
+	closedir(dir);
+	if (!*num) {
+		infof("No suitable input events devices found.");
+		return -1;
+	}
+
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+sysfs_write_int(const char *fname, int val)
+{
+	FILE *f;
+
+	if (!(f = fopen(fname, "w"))) {
+		errorf("Can't open \"%s\" for writing", fname);
+		return -1;
+	}
+
+	fprintf(f, "%d\n", val);
+	fclose(f);
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int is_display_on = 1;
```

Comment:
Unless you know for sure that display is initially on (or off), you might want to use unset|on|off kind of three state variable instead of boolean -> so that the first on/off would always write to sysfs.